metricq_wizard_backend/api/views/explorer.py

```diff
@@ -0,0 +1,182 @@
+        self.nodes = {}
+        self.edges = {}
+        self.layout = {}
+        self.y_depth = defaultdict(int)
+        self.add_metric(original_metric, "#417534")
+        self.add_layout(original_metric, 0)
+        self.original_metric = original_metric
+
+    def add_metric(self, new_metric, color="#4aba4a"):
+        self.nodes[new_metric] = {
+            "name": new_metric,
+            "type": "circle",
+            "color": color,
+            "direction": "north",
+        }
+
+    def add_agent(self, new_agent):
+        if new_agent.endswith("combinator"):
+            color = "#55b1e3"
+        elif new_agent.endswith("aggregator"):
+            color = "#5b90bf"
+        else:
+            color = "#3abec8"
+        self.nodes[new_agent] = {
+            "name": new_agent,
+            "type": "rect",
+            "color": color,
+            "direction": "south",
+        }
+
+    def add_edge(self, source, target):
+        self.edges[source + "to" + target] = {"source": source, "target": target}
+
+    def add_layout(self, target, x_depth):
+        self.layout[target] = {"x": x_depth * 75, "y": self.y_depth[x_depth] * 75}
+        self.y_depth[x_depth] += 1
+
+    def insert_metric(self, metric: str, token: str, x_depth: int):
+        if metric not in self.nodes:
+            self.add_metric(metric)
+            self.add_layout(metric, x_depth)
+        self.add_edge(token, metric)
+
+    async def search(self):
+        await self.search_backwards(self.original_metric, 0)
+        await self.search_forwards(self.original_metric, 0)
+
+        return {"nodes": self.nodes, "edges": self.edges, "layout": self.layout}
+
+    async def search_backwards(self, metric, x_depth):
+        metadata_dict = await self.configurator.fetch_metadata([metric])
+        metadata = metadata_dict[next(iter(metadata_dict))]
+        metric_id = metadata["_id"]
+        source_token = metadata["source"]
+        if source_token not in self.nodes:
+            self.add_agent(source_token)
+            self.add_layout(source_token, x_depth - 1)
+
+        self.add_edge(source_token, metric_id)
+
+        if source_token.startswith("transformer"):
+            if source_token.endswith("combinator"):
+                await self.search_combinator_backwards(metric, source_token, x_depth)
+            elif source_token.endswith("aggregator"):
+                await self.search_aggregator_backwards(
+                    metadata["primary"], source_token, x_depth
+                )
+
+    async def search_combinator_backwards(self, metric: str, token: str, x_depth: int):
+        config = (await self.configurator.get_configs([token]))[token]["metrics"][
+            metric
+        ]
+
+        for input in self.parse_combinator_expression(config["expression"]):
+            try:
+                if input not in self.nodes:
+                    self.add_metric(input)
+                    await self.search_backwards(input, x_depth - 2)
```

Comment:
To make it more obvious where the "magic" constant comes from, subtract 1 here and 1 in the function call.
```suggestion
                    await self.search_backwards(input, x_depth - 1)
```